### PR TITLE
Per line history

### PR DESCRIPTION
**Pull request by samoylovfp**
This contains a lot of additional logic

### CODE REVIEW

File: src/resample.rs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            assert_eq!(resample(vec![0.0, 1.0], 5), vec![0.0, 0.25, 0.5, 0.75, 1.0]);
+        }
+
+        #[test]
+        fn complicated_frac() {
+            assert_eq!(
+                resample(vec![0.0, 1.0, 2.0, 3.0], 14),
+                vec![
+                    0.0, 0.23076923, 0.46153846, 0.6923077, 0.9230769, 1.3076923, 1.7692308,
+                    2.2307692, 2.6923077, 2.2307692, 2.923077, 3.6153846, 4.3076925, 3.0
```
Comment (samoylovfp) [inline src/resample.rs]: This is not correct

File: src/bin/ramp_load.rs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+fn main() {
+    let step_width = Duration::from_millis(200);
+    let cpu_num = 24;
+    println!(
+        "Gradually increasing and then decreasing CPU load, period is {:?}",
+        step_width * cpu_num * 2
+    );
+    loop {
+        for i in 0..cpu_num {
+            let load_duration = step_width.mul_f32((cpu_num - i) as f32 * 2.0);
```
Comment (samoylovfp) [inline src/bin/ramp_load.rs]: This is actually an integer multiplication

File: src/resample.rs (hunk 1 above)
Comment (samoylovfp) [reply src/resample.rs]: Fixed

File: src/bin/ramp_load.rs (hunk 2 above)
Comment (samoylovfp) [reply src/bin/ramp_load.rs]: Fixed

**Comment by samoylovfp** [comment]
image

File: src/bin/ramp_load.rs [hunk 3]
```diff
@@ -0,0 +1,32 @@
+use std::time::{Duration, Instant};
+
+/// Generates linearly increasing and then decreasing load
+/// Cpu load graph is supposed to look like a step-wise approximation of a triangle wave
+fn main() {
+    let step_width = Duration::from_millis(200);
+    let cpu_num = 24;
```
Comment (demoth) [inline src/bin/ramp_load.rs]: you can get it from sysinfo

File: src/resample.rs [hunk 4]
```diff
@@ -0,0 +1,128 @@
+pub fn resample(source: Vec<f32>, expected_len: usize) -> Vec<f32> {
+    assert!(expected_len > 0);
+    assert!(!source.is_empty());
+
+    match source.len().cmp(&expected_len) {
+        // No work required
+        std::cmp::Ordering::Equal => source,
+
+        // Interpolate
+        std::cmp::Ordering::Less => {
```
Comment (demoth) [inline src/resample.rs]: can you please comment this block?

File: src/resample.rs [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+                let value = source[left_index] * (1.0 - lerp_percent)
+                    + source.get(left_index + 1).copied().unwrap_or(0.0) * lerp_percent;
+                source_index += step;
+                result.push(value);
+            }
+            result
+        }
+
+        // Downsample
+        std::cmp::Ordering::Greater => {
```
Comment (demoth) [inline src/resample.rs]: can you please comment this block?

File: src/resample.rs [hunk 6]
```diff
@@ -0,0 +1,115 @@
+pub fn resample(source: Vec<f32>, expected_len: usize) -> Vec<f32> {
+    assert!(expected_len > 0);
+    assert!(!source.is_empty());
+
+    match source.len().cmp(&expected_len) {
+        // No work required
+        std::cmp::Ordering::Equal => source,
```
Comment (demoth) [inline src/resample.rs:7]: can it be `::More`?

**Comment by samoylovfp** [comment]
I'll try to improve it tomorrow

File: src/main.rs [hunk 7]
```diff
@@ -56,20 +61,33 @@ fn print_colored(
...
+        *last_load = *cpu_load_samples.last().unwrap();
+        let load_per_char = resample(cpu_load_samples, chars.len());
+
+        // FIXME: spaces still lose information, is it ok?
+        for (cpu_load, c) in load_per_char.into_iter().zip(chars) {
+            let color = get_color(cpu_load / 100.0);
+            stdout.set_color(ColorSpec::new().set_fg(Some(color)))?;
+            stdout.write_all(c.to_string().as_bytes())?;
+        }
+    } else {
```
Comment (demoth) [inline src/main.rs:86]: else if cpu_load_samples.len() == 1 ?
else if cpu_loads.is_empty ?

File: src/bin/ramp_load.rs (hunk 3 above)
Comment (samoylovfp) [reply src/bin/ramp_load.rs]: Got it from there

File: src/resample.rs (hunk 4 above)
Comment (samoylovfp) [reply src/resample.rs]: Done

File: src/resample.rs (hunk 5 above)
Comment (samoylovfp) [reply src/resample.rs]: Done

File: src/main.rs (hunk 7 above)
Comment (samoylovfp) [reply src/main.rs:86]: cannot be empty, it is created with a single element on line 69.
If there is just a single element, then no processing is required

File: src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,3 @@
+pub mod resample_tools;
```
Comment (demoth) [inline src/lib.rs:1]: what this files does?

**Review by demoth — APPROVED** [review]

File: src/lib.rs (hunk 8 above)
Comment (samoylovfp) [reply src/lib.rs:1]: It tells the rust compiler to create a module called "resample_tools" from `src/resample_tools.rs` available from this library crate

Comment (samoylovfp) [reply src/lib.rs:1]: Which is used for example in this doctest
https://github.com/demoth/cpucat/pull/3/files#diff-385473d0d79ed1ce67e90fea85b9f990680beefd4aa9bd3a2ba38b81e6bff1abR29-R37

Comment (samoylovfp) [reply src/lib.rs:1]: Ah you mean the `lib.rs`? I tried to explain in the commit message. https://github.com/demoth/cpucat/pull/3/commits/8727c205e53265a3631cf1383eef568dba86f104

**Pull request merged**